src/scripts/kube-jobs/recencyToBreachId.js

```diff
@@ -0,0 +1,160 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+ * Executes once
+ * Execute after `deprecateBreachResolve.js`
+ * The purpose of the script is to convert all recency indices to breach ids for `breach_resolution`
+ * For backwards compatibility, all converted `breach_resolution` fields will have a boolean
+ * `useBreachId: true/false`
+ */
+
+import Knex from 'knex'
+import knexConfig from '../../db/knexfile.js'
+import { getAllBreachesFromDb } from '../../utils/hibp.js'
+import { getAllEmailsAndBreaches } from '../../utils/breaches.js'
+const knex = Knex(knexConfig)
+
+const LIMIT = 1000 // with millions of records, we have to load a few at a time
+let subscribersArr = []
+
+const selectAndLockResolutions = async () => {
+  const trx = await knex.transaction()
+  let subscribersArr = []
+  try {
+    subscribersArr = await knex.select('id', 'primary_email', 'breach_resolution')
+      .from('subscribers')
+      .whereNotNull('breach_resolution')
+      .whereNull('db_migration_2')
+      .limit(LIMIT)
+      .orderBy('updated_at', 'desc')
+      .transacting(trx)
+      .forUpdate()
+
+    // update the lock
+    await Promise.all(subscribersArr.map(sub => {
+      const { id } = sub
+      return knex('subscribers')
+        .where('id', id)
+        .update({
+          db_migration_2: true
+        })
+        .transacting(trx)
+    }))
+
+    await trx.commit()
+  } catch (error) {
+    await trx.rollback()
+    console.log('select & mark rows failed!! first row:')
+    console.log({ first: subscribersArr[0] })
+    console.error(error)
+  }
+
+  return subscribersArr
+}
+
+/**
+ * Batch update
+ *
+ * @param {*} updateCollection
+ */
+const batchUpdate = async (updateCollection) => {
+  const trx = await knex.transaction()
+  try {
+    await Promise.all(updateCollection.map(tuple => {
+      const { user, updatedBreachesResolution } = tuple
+      return knex('subscribers')
+        .where('id', user.id)
+        .update({
+          breach_resolution: updatedBreachesResolution
+        })
+        .transacting(trx)
+    }))
+    await trx.commit()
+  } catch (error) {
+    await trx.rollback()
+    console.error('batch update failed!!')
+    console.log({ updateCollection })
+    console.error(error)
+  }
+}
+
+// Script begins here
+const startTime = Date.now()
+console.log(`Start time is: ${startTime}`)
+
+// load all breaches for ref
+const allBreaches = await getAllBreachesFromDb()
+if (allBreaches && allBreaches.length > 0) console.log('breaches loaded successfully! ', allBreaches.length)
+// console.log(JSON.stringify(allBreaches[0]))
+
+// find all subscribers who resolved any breaches in the past,
+// replace recency index with breach id
+
+let failedToSelect = true
+while (failedToSelect) {
+  try {
+    subscribersArr = await selectAndLockResolutions()
+    failedToSelect = false
+  } catch (e) {
+    failedToSelect = true
```

Comment:
I think this is already set to `true` above on L95.
But does this need some sort of sleep or backoff or total number of attempts before launching itself into the sun?
Not sure how possible it is that `selectAndLockResolutions()` might fail and put this in an infinite loop.

Reply:
Fixed, thanks

Reply:
The select is expected to return without error even if we run out of records. The only way it should error out is if the update (lock marking) failed due to race condition. We haven't seen this happen during our tests due to these kubernetes pods being spun up with a slight delay